src/java.base/share/classes/java/io/SerializationMisdeclarationChecker.java

```diff
@@ -0,0 +1,300 @@
+        }
+    }
+
+    private void privilegedCheckSerialPersistentFields() {
+        Field f = declaredField(cl, SERIAL_PERSISTENT_FIELDS_NAME);
+        if (f == null) {
+            return;
+        }
+        if (cl.isRecord()) {
+            commitEvent(SER_PERS_INEFFECTIVE_RECORD,
+                    SERIAL_PERSISTENT_FIELDS_NAME + " is not effective in a record class");
+        } else if (cl.isEnum()) {
+            commitEvent(SER_PERS_INEFFECTIVE_ENUM,
+                    SERIAL_PERSISTENT_FIELDS_NAME + " is not effective in an enum class");
+        }
+        if (!isPrivate(f)) {
+            commitEvent(SER_PERS_PRIVATE,
+                    SERIAL_PERSISTENT_FIELDS_NAME + " must be declared private to be effective");
+        }
+        if (!isStatic(f)) {
+            commitEvent(SER_PERS_STATIC,
+                    SERIAL_PERSISTENT_FIELDS_NAME + " must be declared static to be effective");
+        }
+        if (!isFinal(f)) {
+            commitEvent(SER_PERS_FINAL,
+                    SERIAL_PERSISTENT_FIELDS_NAME + " must be declared final to be effective");
+        }
+        if (f.getType() != ObjectStreamField[].class) {
+            commitEvent(SER_PERS_TYPE_OSF_ARRAY,
+                    SERIAL_PERSISTENT_FIELDS_NAME + " should be declared of type ObjectStreamField[]");
+        }
+        if (!isStatic(f)) {
+            return;
+        }
+        f.setAccessible(true);
+        Object spf = getObject(f);
+        if (spf == null) {
+            commitEvent(SER_PERS_NOT_NULL,
+                    SERIAL_PERSISTENT_FIELDS_NAME + " must not be null to be effective");
+            return;
+        }
+        if (!(spf instanceof ObjectStreamField[])) {
+            commitEvent(SER_PERS_VALUE_OSF_ARRAY,
+                    SERIAL_PERSISTENT_FIELDS_NAME + " must be an instance of ObjectStreamField[] to be effective");
+        }
+    }
+
+    private void privilegedCheckPrivateMethod(String name, Class<?>[] paramTypes, Class<?> retType) {
+        for (Method m : cl.getDeclaredMethods()) {
+            if (m.getName().equals(name)) {
+                privilegedCheckPrivateMethod(m, paramTypes, retType);
+            }
+        }
+    }
+
+    private void privilegedCheckPrivateMethod(Method m, Class<?>[] paramTypes, Class<?> retType) {
+        if (cl.isEnum()) {
+            commitEvent(PRIV_METH_INEFFECTIVE_ENUM,
+                    m + " is not effective on an enum class");
+        } else if (cl.isRecord()) {
+            commitEvent(PRIV_METH_INEFFECTIVE_RECORD,
+                    m + " is not effective on a record class");
+        }
+        if (!isPrivate(m)) {
+            commitEvent(PRIV_METH_PRIV,
+                    m + " must be private to be effective");
+        }
+        if (isStatic(m)) {
+            commitEvent(PRIV_METH_NON_STATIC,
+                    m + " must be non-static to be effective");
+        }
```

Comment:
Should there also be a check to see if this `private` method is (misdeclared) as a `native` method?

Reply:
I'm not sure that a `native` method is not considered effective by serialization. I have to check.

Reply:
@RogerRiggs The checks are agnostic about a method being `native` or not, so I'm not sure I get your point about `native` methods.

Reply:
Right, there's nothing to change; a method declared as native is not mis-declared.